03-astropy-overview/io_fits_ascii_overview.ipynb

```diff
@@ -432,37 +316,53 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "from astropy.table import QTable\n",
-    "from astropy.utils.data import download_file"
+    "mask_hdu.header"
    ]
   },
   {
-   "cell_type": "code",
-   "execution_count": null,
+   "cell_type": "markdown",
    "metadata": {},
-   "outputs": [],
    "source": [
-    "table_filename = download_file('http://data.astropy.org/tutorials/FITS-tables/chandra_events.fits', cache=True)"
+    "You might expect that you can use the `.data` attribute to access the data, but that does not work:"
    ]
   },
   {
    "cell_type": "code",
    "execution_count": null,
-   "metadata": {},
+   "metadata": {
+    "editable": true,
+    "slideshow": {
+     "slide_type": ""
+    },
+    "tags": [
+     "raises-exception"
+    ]
+   },
    "outputs": [],
    "source": [
-    "# When a table contains columns with units, QTable can be used\n",
-    "# instead of Table.\n",
-    "tab = QTable.read(table_filename, hdu=1)"
+    "mask_hdu.data"
    ]
   },
   {
-   "cell_type": "code",
-   "execution_count": null,
+   "cell_type": "markdown",
    "metadata": {},
-   "outputs": [],
    "source": [
-    "tab.colnames"
+    "#### Explanation\n",
+    "\n",
+    "What just happened?! It certainly appeared as if the mask HDU had been loaded since we could access its header, but the data clearly was not loaded. \n",
+    "\n",
+    "The error message, `alueError: I/O operation on closed file`, provides a clue.\n",
+    "\n",
+    "The reason is primarily historical. FITS files are, by default, opened in *memory map* mode to avoid reading in the entire file at once. In memory map mode, the `.data` that looks like an array is actually, because of the way numpy handles memory maps, a refernce to the file with the data. That has been the default in astropy since its very beginning, including in the stand-alone `pyfits` package that preceded, and then became a part of, astropy. \n",
+    "\n",
+    "\n",
+    "On the other hand, a context manager is supposed to guarantee that any resources created in the context, like opening a file handle, is cleaned up, i.e. removed, when you exit the context.\n",
+    "\n",
+    "That has led to a compromise:\n",
+    "\n",
+    "1. If no reference is made to the data in the with block, e.g. by assigning the data to a variable, and the file is opened with `memmap=True`, the default, then all file handles are closed when you exit the block and you have no access to the data after exiting the block. You do have access to the header because that is always read from disk. This is the case we saw above.\n",
```

Comment:
```suggestion
    "1. If no reference is made to the data in the with block, e.g., by assigning the data to a variable, and the file is opened with `memmap=True`, the default, then all file handles are closed when you exit the block and you have no access to the data after exiting the block. You do have access to the header because that is always read from disk. This is the case we saw above.\n",
```